pkg/cache/tas_flavor_snapshot.go

```diff
@@ -0,0 +1,349 @@
+		lowerFitDomains := s.lowerLevelDomains(levelIdx, currFitDomain)
+		sortedLowerDomains := s.sortedDomains(lowerFitDomains)
+		currFitDomain = s.updateCountsToMinimum(sortedLowerDomains, count)
+	}
+	return s.buildAssignment(currFitDomain)
+}
+
+func (s *TASFlavorSnapshot) resolveLevelIdx(
+	topologyRequest *kueue.PodSetTopologyRequest) (int, bool) {
+	var levelKey string
+	if topologyRequest.Required != nil {
+		levelKey = *topologyRequest.Required
+	} else if topologyRequest.Preferred != nil {
+		levelKey = *topologyRequest.Preferred
+	}
+	levelIdx := slices.Index(s.levelKeys, levelKey)
+	if levelIdx == -1 {
+		return levelIdx, false
+	}
+	return levelIdx, true
+}
+
+func (s *TASFlavorSnapshot) findLevelWithFitDomains(levelIdx int, required bool, count int32) (int, []*domain) {
+	levelDomains := s.domainsForLevel(levelIdx)
+	if len(levelDomains) == 0 {
+		return 0, nil
+	}
+	sortedDomain := s.sortedDomains(levelDomains)
+	topDomain := sortedDomain[0]
+	if s.state[topDomain.id] < count {
+		if required {
+			return 0, nil
+		}
+		if levelIdx > 0 {
+			return s.findLevelWithFitDomains(levelIdx-1, required, count)
+		}
+		lastIdx := 0
+		remainingCount := count - s.state[sortedDomain[lastIdx].id]
+		for remainingCount > 0 && lastIdx < len(sortedDomain)-1 {
+			lastIdx++
+			remainingCount -= s.state[sortedDomain[lastIdx].id]
+		}
+		if remainingCount > 0 {
+			return 0, nil
+		}
+		return 0, sortedDomain[:lastIdx+1]
+	}
+	return levelIdx, []*domain{topDomain}
+}
+
+func (s *TASFlavorSnapshot) updateCountsToMinimum(domains []*domain, count int32) []*domain {
+	result := make([]*domain, 0)
+	remainingCount := count
+	for i := 0; i < len(domains); i++ {
+		domain := domains[i]
+		if s.state[domain.id] >= remainingCount {
+			s.state[domain.id] = remainingCount
+			result = append(result, domain)
+			return result
+		}
+		if s.state[domain.id] > 0 {
+			remainingCount -= s.state[domain.id]
+			result = append(result, domain)
+		}
+	}
+	s.log.Error(errCodeAssumptionsViolated, "unexpected remainingCount",
+		"remainingCount", remainingCount,
+		"count", count,
+		"levelValuesPerDomain", s.levelValuesPerDomain,
+		"freeCapacityPerDomain", s.freeCapacityPerDomain)
+	return nil
+}
+
+func (s *TASFlavorSnapshot) buildAssignment(domains []*domain) *kueue.TopologyAssignment {
+	assignment := kueue.TopologyAssignment{
+		Levels:  s.levelKeys,
+		Domains: make([]kueue.TopologyDomainAssignment, 0),
+	}
+	for i := 0; i < len(domains); i++ {
+		assignment.Domains = append(assignment.Domains, kueue.TopologyDomainAssignment{
+			Values: s.asLevelValues(domains[i].id),
+			Count:  s.state[domains[i].id],
+		})
+	}
+	return &assignment
+}
+
+func (s *TASFlavorSnapshot) asLevelValues(domainID utiltas.TopologyDomainID) []string {
+	result := make([]string, len(s.levelKeys))
+	for i := range s.levelKeys {
+		result[i] = s.levelValuesPerDomain[domainID][i]
+	}
+	return result
+}
+
+func (s *TASFlavorSnapshot) lowerLevelDomains(levelIdx int, infos []*domain) []*domain {
```

Comment:
Can we rename `infos` to `domains`

Reply:
Done in the follow up https://github.com/kubernetes-sigs/kueue/pull/3291